### PR TITLE
[MCOL-4991] Solving TRUNCATE/ROUND/CEILING functions on TIME/DATETIME/TIMESTAMP

### PR DESCRIPTION
**Pull request by MuHe03**
Solved the metioned issues on TIME, DATETIME and TIMESTAMP datatypes executions in TRUNCATE/ROUND/CEILING functions.

### CODE REVIEW

File: utils/funcexp/func_round.cpp [hunk 1]
```diff
@@ -734,13 +664,109 @@ string Func_round::getStrVal(Row& row, FunctionParm& parm, bool& isNull, Calpont
...
                                       CalpontSystemCatalog::ColType& op_ct)
 {
-  return parm[0]->data()->getIntVal(row, isNull);
+  int32_t s = parm.size() > 1 ? parm[1]->data()->getIntVal(row, isNull) : 0;
+  if (isNull)
+    return 0;
+  s = (s > 6) ? 6 : s;
+  if (s < 0) 
+  {
+    // int64_t x = parm[0]->data()->getDatetimeIntVal(row, isNull);
```
Comment (drrtuy) [inline utils/funcexp/func_round.cpp]: It is good to clean the commented text up.

File: utils/funcexp/func_round.cpp [hunk 2]
```diff
@@ -734,13 +664,109 @@ string Func_round::getStrVal(Row& row, FunctionParm& parm, bool& isNull, Calpont
 int64_t Func_round::getDatetimeIntVal(Row& row, FunctionParm& parm, bool& isNull,
                                       CalpontSystemCatalog::ColType& op_ct)
 {
-  return parm[0]->data()->getIntVal(row, isNull);
+  int32_t s = parm.size() > 1 ? parm[1]->data()->getIntVal(row, isNull) : 0;
+  if (isNull)
+    return 0;
+  s = (s > 6) ? 6 : s;
```
Comment (drrtuy) [inline utils/funcexp/func_round.cpp]: Could you make this 6 a constexpr const int32_t constant with a meaningful name?

Comment (drrtuy) [reply utils/funcexp/func_round.cpp]: It would be easier to read/maintain.

File: utils/funcexp/func_round.cpp [hunk 3]
```diff
@@ -734,13 +664,109 @@ string Func_round::getStrVal(Row& row, FunctionParm& parm, bool& isNull, Calpont
...
+    //   if (x % 10 >= 5)
+    //     x += 10;
+    //   x *= helpers::powerOf10_c[s];
+    // }
+    // return dataconvert::DataConvert::intToDatetime(x);
+    s = 0;
+  }
+  // else 
+  // {
+    int64_t x = parm[0]->data()->getDatetimeIntVal(row, isNull) + (s <= 5 ? 5 * helpers::powerOf10_c[5 - s] : 0) + (s == 0 ? 48576 : 0); // 48576 = 0xfffff - 1000000;
```
Comment (drrtuy) [inline utils/funcexp/func_round.cpp]: Technically 5 - s can be an arbitrary number and there will be a SEGV called accessing, let's say powerOf10_c[2**31-1] so it is worth to add add a value range check here and everywhere else.

**Review by drrtuy — CHANGES_REQUESTED** [review]
Thank you very much for the bug fix @MuHe03. We appreciate you contribution. I like the patch in general. There are some bits that will make the patch even better. 
There is a lot of boilerplate code so I suggest you to produce a inline pure function that calculates the value. It should be reused in the actual compilation units, e.g. func_ceil.cpp etc.
It is up to you to have a single pure func or func per compilation unit.

File: utils/funcexp/func_round.cpp (hunk 1 above)
Comment (MuHe03) [reply utils/funcexp/func_round.cpp]: Sure, but I'm still not sure, whether it's necessary to do truncate and round before the decimal point, which is not supported in the InnoDB. If it's not, then I will remove the commented part, which I wrote to support this.

File: utils/funcexp/func_round.cpp (hunk 3 above)
Comment (MuHe03) [reply utils/funcexp/func_round.cpp]: It's not necessary, because I already make sure that s is between 0 and 6 which won't cause out of index.

**Review by drrtuy — APPROVED** [review]

**Review by mariadb-AndreyPiskunov — APPROVED** [review]

**Review by drrtuy — CHANGES_REQUESTED** [review]
Could you take a look at the MTR test failed.

**Review by drrtuy — APPROVED** [review]

**Pull request merged**